Mathlib/GroupTheory/PushoutI.lean

```diff
@@ -0,0 +1,657 @@
+of the amalgamated product of groups (i.e. the special case where all the maps
+in the diagram are injective).
+
+## Main definitions
+
+- `Monoid.PushoutI`: the pushout of a diagram of monoids indexed by a type `ι`
+- `Monoid.PushoutI.base`: the map from the amalgamating monoid to the pushout
+- `Monoid.PushoutI.of`: the map from each Monoid in the family to the pushout
+- `Monoid.PushoutI.lift`: the universal property used to define homomorphisms out of the pushout.
+
+- `Monoid.PushoutI.NormalWord`: a normal form for words in the pushout
+- `Monoid.PushoutI.of_injective`: if all the maps in the diagram are injective in a pushout of
+groups then so is `of`
+- `Monoid.PushoutI.Reduced.eq_empty_of_mem_range`: For any word `w` in the coproduct,
+if `w` is reduced (i.e none its letters are in the image of the base monoid), and nonempty, then
+`w` itself is not in the image of the base monoid.
+
+## References
+
+* The normal form theorem follows these [notes](https://webspace.maths.qmul.ac.uk/i.m.chiswell/ggt/lecture_notes/lecture2.pdf)
+from Queen Mary University
+
+## Tags
+
+amalgamated product, pushout, group
+
+-/
+
+namespace Monoid
+
+open CoprodI Subgroup Coprod Function List
+
+variable {ι : Type*} {G : ι → Type*} {H : Type*} {K : Type*} [Monoid K]
+
+/-- The relation we quotient by to form the pushout -/
+def PushoutI.con [∀ i, Monoid (G i)] [Monoid H] (φ : ∀ i, H →* G i) :
+    Con (Coprod (CoprodI G) H) :=
+  conGen (fun x y : Coprod (CoprodI G) H =>
+    ∃ i x', x = inl (of (φ i x')) ∧ y = inr x')
+
+/-- The indexed pushout of monoids, which is the pushout in the category of monoids,
+or the category of groups. -/
+def PushoutI [∀ i, Monoid (G i)] [Monoid H] (φ : ∀ i, H →* G i) : Type _ :=
+  (PushoutI.con φ).Quotient
+
+namespace PushoutI
+
+section Monoid
+
+variable [∀ i, Monoid (G i)] [Monoid H] {φ : ∀ i, H →* G i}
+
+protected instance mul : Mul (PushoutI φ) := by
+  delta PushoutI; infer_instance
+
+protected instance one : One (PushoutI φ) := by
+  delta PushoutI; infer_instance
+
+instance monoid : Monoid (PushoutI φ) :=
+  { Con.monoid _ with
+    toMul := PushoutI.mul
+    toOne := PushoutI.one }
+
+/-- The map from each indexing group into the pushout -/
+def of (i : ι) : G i →* PushoutI φ :=
+  (Con.mk' _).comp <| inl.comp CoprodI.of
+
+/-- The map from the base monoid into the pushout -/
+def base : H →* PushoutI φ :=
+  (Con.mk' _).comp inr
+
+theorem of_comp_eq_base (i : ι) : (of i).comp (φ i) = (base (φ := φ)) := by
+  ext x
+  apply (Con.eq _).2
+  refine ConGen.Rel.of _ _ ?_
+  simp only [MonoidHom.comp_apply, Set.mem_iUnion, Set.mem_range]
+  exact ⟨_, _, rfl, rfl⟩
+
+theorem of_apply_eq_base (i : ι) (x : H) : of i (φ i x) = base (φ := φ) x := by
+  rw [← MonoidHom.comp_apply, of_comp_eq_base]
+
+/-- Define a homomorphism out of the pushout of monoids be defining it on each object in the
+diagram -/
+def lift (f : ∀ i, G i →* K) (k : H →* K)
```

Comment:
On the other hand `f` and `k` can be implicit here (but it can make things uglier later on).